### ISSUE TITLE
Shortcut not detected

### ISSUE THREAD

**Issue by horta** [knubie/vim-kitty-navigator#2]
Inside Neovim, pressing any of the shortcuts `ctrl+h`, `ctrl+j`, `ctrl+k`, or `ctrl+l` results in

```
Traceback (most recent call last):
  File "kitty/boss.py", line 507, in dispatch_special_key
  File "kitty/boss.py", line 567, in dispatch_action
  File "kitty/boss.py", line 712, in kitten
  File "kitty/boss.py", line 669, in _run_kitten
  File "/Users/horta/.config/kitty/pass_keys.py", line 29, in handle_result
    w.send_text(key)
  File "kitty/window.py", line 245, in send_text
ValueError: not enough values to unpack (expected 2, got 1)
```

Adding a `print(args)` statement right before before `mods, key, is_text = ku.parse_kittens_shortcut(args[3])` prints

```
['pass_keys.py', 'neighboring_window', 'right', 'ctrl+l']
```

When I press `ctrl+l` from Neovim.

**Comment by horta** [comment]
Kitty version: 0.13.3

**Comment by taylorthurlow** [comment]
I'm experiencing the same issue. It happens when using a navigation shortcut when inside Neovim.

**Comment by taylorthurlow** [comment]
Did some more digging, still don't know python, haha.

It seems like `parse_kittens_shortcut` is giving weird values. I'd expect when you pass it a value of `ctrl+j` that it would give you `mods = 2`, `key = 'j'`, and `is_text = False`. Mods being 2 because SHIFT has a value of 1, CONTROL a value of 2, ALT a value of 4, and SUPER a value of 8. This way the addition of these values tells us what modifiers are pressed. That being said, `parse_kittens_shortcut` `mods` value is `4`, which says that it thinks the ALT modifier is being pressed instead. On top of that, `is_text` is actually set to `True`, so the modifier is ignored completely and the character `j` gets sent to the terminal and it just moves my cursor down a line as normal. Also, it seems like the key itself should be capitalized, because non-capitalized key names breaks the `getattr(keys.defines, 'GLFW_KEY_{}'.format(key))` call.

If I force `mods` to be equal to `2` (for CONTROL), force `is_text` to be `False`, and capitalize the key character, it doesn't have an error, but it's translating my `ctrl+j` shortcut into a newline (`\n`) character. I suppose this has something to do with the fact that CTRL+J is [historically a newline button?](https://vi.stackexchange.com/a/4249/21072)

---

So I didn't really achieve anything, but my unprofessional opinion is that something is fucky with `parse_kittens_shortcut`.

**Comment by knubie** [comment]
Hey @horta @taylorthurlow sorry for the late response. Somehow I missed this in my inbox.

To be honest I'm not really sure what's going on here either, and I copied a lot of the code in `pass_keys.py` from this other [kitten](https://github.com/yurikhan/kitty-smart-scroll). So I don't really have a good grasp on a lot of the stuff that's happening below line 26. I feel like there must be a simpler way to pass keys through, but 🤷‍♂️ .

I'd be happy to take a PR if either of you can figure out what's wrong.

**Comment by taylorthurlow** [comment]
Cool, I'll see what I can do.

**Comment by horta** [comment]
I found a workaround. Add ``getattr(tab, args[1])(args[2])`` before line ``if w.screen.is_main_linebuf():``. I guess ``is_main_linebuf`` is not working as expected.

**Comment by taylorthurlow** [comment]
Looks like that sort of works, I'm having an issue where it's skipping neovim buffers. Consider the following window:

![Screen Shot 2019-03-14 at 1 02 05 PM](https://user-images.githubusercontent.com/761640/54387753-65796900-4659-11e9-8931-eb0f678d0c3f.png)

No matter which neovim buffer I'm in, if I press `ctrl+k` it moves me to the kitty window above neovim. If I press `ctrl+j` while in the upper kitty window, it *sometimes* skips the uppermost neovim buffer.

**Comment by taylorthurlow** [comment]
I restarted Kitty (thought I had) and I think the skipping pane problem went away, but navigating within neovim still skips buffers and goes right to the neighboring kitty window.

**Comment by taylorthurlow** [comment]
Alright, I solved my problem. I had the kitten scripts in the wrong place so they didn't work when called from the vimscript, so that's what caused my previous issue. Here's my revised version of `pass_keys.py`:

~~~python
import kitty.conf.utils as ku
import kitty.key_encoding as ke
from kitty import keys

def main():
    pass


def actions(extended):
    yield keys.defines.GLFW_PRESS
    if extended:
        yield keys.defines.GLFW_RELEASE


def handle_result(args, result, target_window_id, boss):
    w = boss.window_id_map.get(target_window_id)
    tab = boss.active_tab
    if w is None:
        return

    if w.screen.is_main_linebuf():
        getattr(tab, args[1])(args[2])
        return

    mods, key, is_text = ku.parse_kittens_shortcut(args[3])
    control_characters = {
        'h': '\x08',
        'j': '\x0A',
        'k': '\x0B',
        'l': '\x0C',
    }
    w.send_text('application,normal', control_characters[key])


handle_result.no_ui = True
~~~

I left the `is_main_linebuf` alone, I'm pretty sure it's working properly. I ripped out the extra text processing, because we know if this script is handling a shortcut, it has to be one of the `ctrl+/h/j/k/l` keys. I hard-coded the [values for the control characters](http://jkorpela.fi/chars/c0.html) and pass them directly. This means that the shortcuts aren't customizable anymore, if you can find a way to build the hex code for the desired key, then go for it, but it was more work than I was willing to put in. I added `'application,normal'` as the mode for `send_text` to avoid errors, this seems to work just fine.

@knubie I'm willing to submit a PR, but seeing as this breaks customizing the shortcut itself, I haven't.

**Comment by horta** [comment]
I still had to add ``getattr(tab, args[1])(args[2])`` before the line ``if w.screen.is_main_linebuf():``, @taylorthurlow , otherwise the cursor won't leave neovim. And yes, I have the same bug you saw: neovim skips its own windows and goes directly to kitty window. I believe this is a problem with the vim script instead.

**Comment by taylorthurlow** [comment]
@horta I was in the exact same position you're in. The skipping problem is due to you placing `getattr` where you did. I started investigating the vim script, so play around with sending commands to kitty the same way the vimscript does. I have mine working 100% without adding `getattr` and without modifying the vimscript.

**Comment by knubie** [comment]
The only problem I've had with `w.screen.is_main_linebuf()` is that it will return false if any applications is running, not just neovim. For example if I have [tig](https://github.com/jonas/tig) open, it will send the shortcuts to it.

I got some feedback from kovidgoyal in this [thread](https://github.com/kovidgoyal/kitty/issues/1254) on how to make that conditional more specific, but haven't gotten around to implementing it.

**Comment by aymanbagabas** [comment]
Any updates?

**Comment by horta** [comment]
I have been using neovim's terminal (and neovim splits) and kitty tabs. Kitty splits became obsolete to me.

**Comment by lovetocode999** [comment]
I opened a PR that attempts to fix this issue [here](https://github.com/knubie/vim-kitty-navigator/pull/6)

**Comment by knubie** [comment]
Closed by 56c70a95d8db84b76083f71633a5eeb5698fbf49 and ebdf6988bac3d1ebc45b3e51a86f3c065cf60c24
